test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -0,0 +1,226 @@
+// Those variables should have corresponding patches which set the etcd and CoreDNS tags in KCP.
+func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() clusterUpgradeWithRuntimeSDKSpecInput) {
+	const (
+		textExtensionPathVariable = "TEST_EXTENSION"
+		specName                  = "k8s-upgrade-with-runtimesdk"
+	)
+
+	var (
+		input         clusterUpgradeWithRuntimeSDKSpecInput
+		namespace     *corev1.Namespace
+		ext           *runtimev1.ExtensionConfig
+		cancelWatches context.CancelFunc
+
+		controlPlaneMachineCount int64
+		workerMachineCount       int64
+
+		clusterResources  *clusterctl.ApplyClusterTemplateAndWaitResult
+		testExtensionPath string
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeFrom))
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersionUpgradeTo))
+		Expect(input.E2EConfig.Variables).To(HaveKey(EtcdVersionUpgradeTo))
+		Expect(input.E2EConfig.Variables).To(HaveKey(CoreDNSVersionUpgradeTo))
+
+		testExtensionPath = input.E2EConfig.GetVariable(textExtensionPathVariable)
+		Expect(testExtensionPath).To(BeAnExistingFile(), "The %s variable should resolve to an existing file", textExtensionPathVariable)
+
+		if input.ControlPlaneMachineCount == nil {
+			controlPlaneMachineCount = 1
+		} else {
+			controlPlaneMachineCount = *input.ControlPlaneMachineCount
+		}
+
+		if input.WorkerMachineCount == nil {
+			workerMachineCount = 2
+		} else {
+			workerMachineCount = *input.WorkerMachineCount
+		}
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the Namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+		clusterResources = new(clusterctl.ApplyClusterTemplateAndWaitResult)
+	})
+
+	It("Should create and upgrade a workload cluster", func() {
+		By("Deploy Test Extension")
+		testExtensionDeploymentTemplate, err := os.ReadFile(testExtensionPath) //nolint:gosec
+		Expect(err).ToNot(HaveOccurred(), "Failed to read the extension config deployment manifest file")
+		testExtensionDeployment := strings.ReplaceAll(string(testExtensionDeploymentTemplate), "${SERVICE_NAMESPACE}", namespace.Name)
```

Comment:
nit: let's document why we have to do this replace

Reply:
might be I'm wrong, but I haven't found where this is used...

Reply:
Good point. I'll document it.

The problem is that the certificate requires dnsNames which contain the namespace. But as I can't render that statically when rendering the test extension deployment.yaml I have to do it here.

(the alternative is looping overall objects like clusterctl and modifying the namespace)